jsapp/js/components/permissions/userAssetPermsEditor.es6

```diff
@@ -0,0 +1,468 @@
+import React from 'react';
+import reactMixin from 'react-mixin';
+import autoBind from 'react-autobind';
+import Reflux from 'reflux';
+import TagsInput from 'react-tagsinput';
+import Checkbox from 'js/components/checkbox';
+import TextBox from 'js/components/textBox';
+import stores from 'js/stores';
+import actions from 'js/actions';
+import bem from 'js/bem';
+import classNames from 'classnames';
+import permParser from './permParser';
+import {
+  assign,
+  t,
+  notify,
+  buildUserUrl
+} from 'js/utils';
+
+/**
+ * Form for adding/changing user permissions for surveys.
+ *
+ * @prop uid - asset uid
+ * @prop username - permissions user username (could be empty for new)
+ * @prop permissions - list of permissions (could be empty for new)
+ * @prop onSubmitEnd - callback to be run when submit ends (success or failure)
+ */
+class UserAssetPermsEditor extends React.Component {
+  constructor(props) {
+    super(props);
+    autoBind(this);
+
+    this.state = {
+      // inner workings
+      usernamesBeingChecked: new Set(),
+      isSubmitPending: false,
+      isEditingUsername: false,
+      // form user inputs
+      username: '',
+      formView: false,
+      formViewDisabled: false,
+      formEdit: false,
+      submissionsView: false,
+      submissionsViewDisabled: false,
+      submissionsViewPartial: false,
+      submissionsViewPartialDisabled: false,
+      submissionsViewPartialUsers: [],
+      submissionsAdd: false,
+      submissionsEdit: false,
+      submissionsEditDisabled: false,
+      submissionsValidate: false,
+      submissionsValidateDisabled: false
+    };
+
+    this.applyPropsData();
+  }
+
+  /**
+   * Fills up form with provided user name and permissions (if applicable)
+   */
+  applyPropsData() {
+    if (this.props.permissions) {
+      const formData = permParser.buildFormData(this.props.permissions);
+      this.state = this.applyValidityRules(assign(this.state, formData));
```

Comment:
If they are invalid in the sense of permissions that are not defined on BE and FE, then they will be completely omitted.

If they are invalid in the sense of BE having contradictory permissions being set, then probably some funky stuff could happen (i.e. FE assumes BE has proper data). For the most part I think it should display a valid set of checked checkboxes as `applyValidityRules` is pretty straightforward on/off function. Also mind that `userAssetPermsEditor` doesn't display permissions per se, it has checkboxes that are somehow mapped to permissions, but are not identical with them.

Wrong permissions being displayed will definitely happen in `jsapp/js/components/permissions/userPermissionRow.es6`, as it just displays the array of permissions from BE.

If you'd like for FE to react somehow to getting invalid BE data, I think it should be checked somewhere in `jsapp/js/actions/permissions.es6` near `permissionsActions.getAssetPermissions` call. Then if data is invalid, I would display error notification to inform that BE has wrong data and to ask for user to either contact admin or remove and add permissions again?

Reply:
As for "Edit submissions" label unfortunately it is hardcoded in FE, as all the other labels in the UPDATE PERMISSIONS form. It has to be this way, because it is not possible to render the permissions 1:1 from BE data due to the partial permissions.

I changed the label to "Change submissions". I could use the values from the BE to some of the labels though - just map the hardcoded codenames from `jsapp/js/constants.es6:19` with the label that I will from BE call. What do you think?

Reply:
> Then if data is invalid, I would display error notification to inform that BE has wrong data and to ask for user to either contact admin or remove and add permissions again?

Yes, this sounds good. Please add something like "The stored permissions are invalid. Please assign them again. If this problem persists, contact help@kobotoolbox.org" (maybe using `replaceSupportEmail()`). I intentionally mangled stuff on the backend as a test; the backend has its own checks against contradictory permissions, so something like this should be rare. It's just a high-stakes kind of rare since there's the potential for unintentional access being granted, so I think having the warning is worthwhile.

> Messing with the applyValidityRules setting and disabling checkboxes would be confusing I think…
OK, let's not do that. The warning alone is fine.

> The real question here is - should FE be prepared for getting wrong data from BE? Because we don't do it anywhere else.

Permissions are more delicate than other things, and determining whether data sent by the BE is actually valid may be difficult or impossible without duplicating lots of code. For less-sensitive places where the BE sends data so bad that the FE cannot render it properly, throwing an exception (that's logged by Sentry) and displaying an error is enough.

> As for "Edit submissions" label unfortunately it is hardcoded in FE, as all the other labels in the UPDATE PERMISSIONS form. It has to be this way, because it is not possible to render the permissions 1:1 from BE data due to the partial permissions.

I'd have to re-read the partial permissions code to fully understand, but let's leave it for a future PR. Thanks for using "Change submissions".

Reply:
Moved to https://github.com/kobotoolbox/kpi/issues/2343